tensorflow/contrib/opt/python/training/irprop_plus.py

```diff
@@ -0,0 +1,319 @@
+    of using this function. The loss argument has to be passed as a 0-D Tensor.
+    Args:
+      loss: A 0-D `Tensor` containing the value to minimize.
+      global_step: Optional `Variable` to increment by one after the
+        variables have been updated.
+      var_list: Optional list or tuple of `Variable` objects to update to
+        minimize `loss`.  Defaults to the list of variables collected in
+        the graph under the key `GraphKeys.TRAINABLE_VARIABLES`.
+      gate_gradients: How to gate the computation of gradients.  Can be
+        `GATE_NONE`, `GATE_OP`, or  `GATE_GRAPH`.
+      aggregation_method: Specifies the method used to combine gradient terms.
+        Valid values are defined in the class `AggregationMethod`.
+      colocate_gradients_with_ops: If True, try colocating gradients with
+        the corresponding op.
+      name: Optional name for the returned operation.
+      grad_loss: Optional. A `Tensor` holding the gradient computed for `loss`.
+    Returns:
+      An Operation that updates the variables in `var_list`.  If `global_step`
+      was not `None`, that operation also increments `global_step`.
+    Raises:
+      ValueError: If some of the variables are not `Variable` objects.
+      ValueError: If the `loss` is not a 0-D tensor (scalar).
+    @compatibility(eager)
+    When eager execution is enabled, `loss` should be a Python function that
+    takes elements of `var_list` as arguments and computes the value to be
+    minimized. If `var_list` is `None`, `loss` should take no arguments.
+    Minimization (and gradient computation) is done with respect to the
+    elements of `var_list` if not `None`, else with respect to any trainable
+    variables created during the execution of the `loss` function.
+    `gate_gradients`, `aggregation_method`, `colocate_gradients_with_ops` and
+    `grad_loss` are ignored when eager execution is enabled.
+    @end_compatibility
+    """
+    # Override method from base class
+
+    # Error E(t)
+    if self._is_scalar(loss):
+      self._error = loss
+    else:
+      raise ValueError(
+          "'loss' (%s) must be a 0-D tensor." % loss)
+
+    return super(IRpropPlusOptimizer, self).minimize(
+        loss,
+        global_step=global_step, var_list=var_list,
+        gate_gradients=gate_gradients,
+        aggregation_method=aggregation_method,
+        colocate_gradients_with_ops=colocate_gradients_with_ops, name=name,
+        grad_loss=grad_loss)
+
+  def _finish(self, update_ops, name_scope):
+
+    # Update the old error E(t-1) <- E(t)
+    with ops.control_dependencies(update_ops):
+      with ops.colocate_with(self._error):
+        old_error_update = self._old_error.assign(
+            self._error, use_locking=self._use_locking)
+
+    return control_flow_ops.group(*update_ops + [old_error_update],
+                                  name=name_scope)
+
+  def apply_gradients(self, grads_and_vars, loss, global_step=None, name=None):
```

Comment:
I don't think you need to overload apply_gradients

Reply:
Same comment as above.